Added filtered index support for SQL Server

cziegenberg opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | #3038

#### Summary

Added support for filtered (or partial) indexes in SQL Server (supported since version 2008). This is the second version after the closed pull request #4240. It contains the implementation for SQL Server only. I will try to add the implementation for SQLite later, in a separate pull request.

The implementation has the following limitation (same as with other platforms like PostgreSQL, see #3871):
The given "where" option may be modified by the server. So when reading it again for comparison, the result may differ, which can result in an unnecessary update when creating a migration. This depends on how the "where" option is defined and especially occurrs in case of unique constraints, because for each indexed column in SQL Server a "IS NOT NULL" filter is added automatically by DBAL (for compatibility with other platforms and the SQL standard). This has to be solved globally (i.e. by saving the original value as a comment or adding a normalization step before the comparison) and is independent of this pull request.

cziegenberg committed:
Added filtered index support for SQL Server

Added support for filtered (or partial) indexes in SQL Server (supported since version 2008).

The implementation has the following limitations, which could be optimized later:
- The given "where" option may be optimized and modified by the server. So when reading it again for comparison, the result may differ, which can result in an update when creating a migration. This limitation is known from other platforms like PostgreSQL and depends on how the "where" option is defined.
- The previous limitation especially occurrs in case of unique constraints, because for each indexed column in SQL Server a "IS NOT NULL" filter is added automatically (for compatibility with other platforms and the SQL standard).

cziegenberg commented:
@morozov As requested, I created an updated pull request for the right branch and with additional tests. Would be great if you could review it once again. Thanks.

morozov commented:
@cziegenberg since `2.11.0` has been released, `2.11.x` only accepts bugfixes and the improvements for the upgrade path to 3.0. Please retarget against `3.0.x` (no need to file another PR, use the Edit button at the top of the page).

morozov commented on tests/Doctrine/Tests/DBAL/Functional/Schema/SQLServerSchemaManagerTest.php:
```diff
@@ -404,4 +405,30 @@ public function testPkOrdering(): void
+
+    public function testPartialIndexes(): void
+    {
+        if (! $this->schemaManager->getDatabasePlatform()->supportsPartialIndexes()) {
```
Would it be possible to move this test one level higher, to the `SchemaManagerFunctionalTestCase` and reuse for all platforms that support partial indexes?

Since it's a functional test, it should cover the API on all platforms that declare support for it, and we already have one in https://github.com/doctrine/dbal/blob/3f3551aadb20c7ed58ee9c7f06ed74770588272f/tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php#L355

cziegenberg replied:
No, not at the moment, because of the limitation/bug #3871 - Postgres and SQL Server automatically add (different) brackets, type hints etc., so the returned where condition depends on the platform (for the same input).

cziegenberg commented:
Sorry, I seems like I did something wrong. I changed the target branch, but could not change the source for this pull request, although I switched the branch on my site...?! So the only thing I could do was to create a new pull request (#4342). So closing this one...

BTW: Good to see, that all the old SQL Server code has been removed now.
